File: imsdia/api/java/test/diameter/s6a/NotifyTest.java
package diameter.s6a;

import com.intellinet.diameter.*;
import diameter.*;

public class NotifyTest {

    public static void testNotifyRequest() throws Exception {
        NotifyRequest notifyRequest = new NotifyRequest(false, false);
        notifyRequest.setSessionId(new diameter.base.SessionId("utf8string-969470417"));
        notifyRequest.setAuthSessionState(new diameter.base.AuthSessionState(diameter.base.AuthSessionState.Value.NO_STATE_MAINTAINED));
        notifyRequest.setOriginHost(new diameter.base.OriginHost("id-573718676.diameter.com"));
        notifyRequest.setOriginRealm(new diameter.base.OriginRealm("id1432601068.diameter.com"));
        notifyRequest.setDestinationRealm(new diameter.base.DestinationRealm("id-1773359194.diameter.com"));
        notifyRequest.setUserName(new diameter.base.UserName("utf8string715447879"));
        {
            diameter.base.VendorSpecificApplicationId vendorSpecificApplicationId = new diameter.base.VendorSpecificApplicationId();
            vendorSpecificApplicationId.addVendorId(new diameter.base.VendorId(new Integer(2127807892)));
            vendorSpecificApplicationId.setAuthApplicationId(new diameter.base.AuthApplicationId(new Integer(1958820869)));
            vendorSpecificApplicationId.setAcctApplicationId(new diameter.base.AcctApplicationId(new Integer(1059138687)));
            notifyRequest.setVendorSpecificApplicationId(vendorSpecificApplicationId);
        }
        notifyRequest.setDestinationHost(new diameter.base.DestinationHost("id910780694.diameter.com"));
        {
            SupportedFeatures supportedFeatures = new SupportedFeatures();
            supportedFeatures.setVendorId(new diameter.base.VendorId(new Integer(-1608486387)));
            supportedFeatures.setFeatureListID(new FeatureListID(new Integer(-456445903)));
            supportedFeatures.setFeatureList(new FeatureList(new Integer(-511470649)));
            // add more AVP as ANY values to supportedFeatures
            notifyRequest.addSupportedFeatures(supportedFeatures);
        }
        {
            TerminalInformation terminalInformation = new TerminalInformation();
            terminalInformation.setIMEI(new IMEI("utf8string1355121194"));
            terminalInformation.setS6a3GPP2MEID(new S6a3GPP2MEID("octetstring-1473450799"));
            terminalInformation.setSoftwareVersion(new SoftwareVersion("utf8string2075961027"));
            // add more AVP as ANY values to terminalInformation
            notifyRequest.setTerminalInformation(terminalInformation);
        }
        notifyRequest.setNORFlags(new NORFlags(new Integer(-1503220117)));
        {
            MIP6AgentInfo mIP6AgentInfo = new MIP6AgentInfo();
            mIP6AgentInfo.addMIPHomeAgentAddress(new MIPHomeAgentAddress(new Address(Address.Family.IPV4, "118.44.6.33")));
            {
                MIPHomeAgentHost mIPHomeAgentHost = new MIPHomeAgentHost();
                mIPHomeAgentHost.setDestinationRealm(new diameter.base.DestinationRealm("id-1162465832.diameter.com"));
                mIPHomeAgentHost.setDestinationHost(new diameter.base.DestinationHost("id-1279728842.diameter.com"));
                // add more AVP as ANY values to mIPHomeAgentHost
                mIP6AgentInfo.setMIPHomeAgentHost(mIPHomeAgentHost);
            }
            mIP6AgentInfo.setMIP6HomeLinkPrefix(new MIP6HomeLinkPrefix("octetstring1368201838"));
            // add more AVP as ANY values to mIP6AgentInfo
            notifyRequest.setMIP6AgentInfo(mIP6AgentInfo);
        }
        notifyRequest.setContextIdentifier(new ContextIdentifier(new Integer(-2014243704)));
        notifyRequest.setServiceSelection(new ServiceSelection("utf8string-203163541"));
        notifyRequest.setAlertReason(new AlertReason(AlertReason.Value.UE_MEMORY_AVAILABLE));
        {
            diameter.base.ProxyInfo proxyInfo = new diameter.base.ProxyInfo();
            proxyInfo.setProxyHost(new diameter.base.ProxyHost("id-1686456816.diameter.com"));
            proxyInfo.setProxyState(new diameter.base.ProxyState("octetstring1255697806"));
            // add more AVP as ANY values to proxyInfo
            notifyRequest.addProxyInfo(proxyInfo);
        }
        notifyRequest.addRouteRecord(new diameter.base.RouteRecord("id-1190118120.diameter.com"));
        // add more AVP as ANY values to notifyRequest

        byte[] buffer = notifyRequest.encode();
        NotifyRequest notifyRequest2 = new NotifyRequest(buffer);
        String s1 = notifyRequest.toString();
        String s2 = notifyRequest2.toString();
        if (!s1.equals(s2)) {
            System.err.println("notifyRequest1 = " + s1);
            System.err.println("notifyRequest2 = " + s2);
            throw new VerificationException("NotifyRequest encoding/decoding failure");
        }
    }

    public static void testNotifyAnswer() throws Exception {
        NotifyAnswer notifyAnswer = new NotifyAnswer(false, false);
        notifyAnswer.setSessionId(new diameter.base.SessionId("utf8string-702101081"));
        notifyAnswer.setAuthSessionState(new diameter.base.AuthSessionState(diameter.base.AuthSessionState.Value.STATE_MAINTAINED));
        notifyAnswer.setOriginHost(new diameter.base.OriginHost("id-951463919.diameter.com"));
        notifyAnswer.setOriginRealm(new diameter.base.OriginRealm("id-1813555853.diameter.com"));
        {
            diameter.base.VendorSpecificApplicationId vendorSpecificApplicationId = new diameter.base.VendorSpecificApplicationId();
            vendorSpecificApplicationId.addVendorId(new diameter.base.VendorId(new Integer(-1954619018)));
            vendorSpecificApplicationId.setAuthApplicationId(new diameter.base.AuthApplicationId(new Integer(-1047677209)));
            vendorSpecificApplicationId.setAcctApplicationId(new diameter.base.AcctApplicationId(new Integer(-1797277240)));
            notifyAnswer.setVendorSpecificApplicationId(vendorSpecificApplicationId);
        }
        notifyAnswer.setResultCode(new diameter.base.ResultCode(new Integer(-242720060)));
        {
            diameter.base.ExperimentalResult experimentalResult = new diameter.base.ExperimentalResult();
            experimentalResult.setVendorId(new diameter.base.VendorId(new Integer(-1351234418)));
            experimentalResult.setExperimentalResultCode(new diameter.base.ExperimentalResultCode(new Integer(-953414410)));
            notifyAnswer.setExperimentalResult(experimentalResult);
        }
        {
            SupportedFeatures supportedFeatures = new SupportedFeatures();
            supportedFeatures.setVendorId(new diameter.base.VendorId(new Integer(-1916652700)));
            supportedFeatures.setFeatureListID(new FeatureListID(new Integer(577144930)));
            supportedFeatures.setFeatureList(new FeatureList(new Integer(-399803244)));
            // add more AVP as ANY values to supportedFeatures
            notifyAnswer.addSupportedFeatures(supportedFeatures);
        }
        {
            diameter.base.ProxyInfo proxyInfo = new diameter.base.ProxyInfo();
            proxyInfo.setProxyHost(new diameter.base.ProxyHost("id412114330.diameter.com"));
            proxyInfo.setProxyState(new diameter.base.ProxyState("octetstring1960140274"));
            // add more AVP as ANY values to proxyInfo
            notifyAnswer.addProxyInfo(proxyInfo);
        }
        notifyAnswer.addRouteRecord(new diameter.base.RouteRecord("id-11339961.diameter.com"));
        // add more AVP as ANY values to notifyAnswer

        byte[] buffer = notifyAnswer.encode();
        NotifyAnswer notifyAnswer2 = new NotifyAnswer(buffer);
        String s1 = notifyAnswer.toString();
        String s2 = notifyAnswer2.toString();
        if (!s1.equals(s2)) {
            System.err.println("notifyAnswer1 = " + s1);
            System.err.println("notifyAnswer2 = " + s2);
            throw new VerificationException("NotifyAnswer encoding/decoding failure");
        }
    }

    public static void main(String[] args) throws Exception {
        testNotifyRequest();
        testNotifyAnswer();
    }
}

File: imsdia/api/c++/mtest/diameter/s6a_r10/src/NotifyTest.cpp
/********************************-*-CPP-*-***********************************
 *                                                                          *
 *     Copyright 2006 IntelliNet Technologies, Inc. All Rights Reserved.    *
 *             Manufactured in the United States of America.                *
 *       1990 W. New Haven Ste. 312, Melbourne, Florida, 32904 U.S.A.       *
 *                                                                          *
 *   This product and related documentation is protected by copyright and   *
 *   distributed under licenses restricting its use, copying, distribution  *
 *   and decompilation.  No part of this product or related documentation   *
 *   may be reproduced in any form by any means without prior written       *
 *   authorization of IntelliNet Technologies and its licensors, if any.    *
 *                                                                          *
 *   RESTRICTED RIGHTS LEGEND:  Use, duplication, or disclosure by the      *
 *   government is subject to restrictions as set forth in subparagraph     *
 *   (c)(1)(ii) of the Rights in Technical Data and Computer Software       *
 *   clause at DFARS 252.227-7013 and FAR 52.227-19.                        *
 *                                                                          *
 ****************************************************************************
 *                                                                          *
 * CONTRACT: INTERNAL                                                       *
 ****************************************************************************
 * LOG: $Log: NotifyTest.cpp,v $
 * LOG: Revision 1.1.6.2  2014/02/12 08:06:13  millayaraja
 * LOG:  added testapp s6a_r10 for IWF
 * LOG:
 * LOG: Revision 1.1.2.2  2013/02/14 09:49:17  brajappa
 * LOG: Notify code added to the app
 * LOG:
 * LOG: Revision 1.1.2.1  2012/12/19 22:17:18  brajappa
 * LOG: Added s6a and s13 Release 10 ABNF
 * LOG:
 * LOG: Revision 1.1.2.1  2012/07/09 08:53:44  pramana
 * LOG: New test code added for S6a, S13 for Release 10 support (Bug 48).
 * LOG:
 * ID: $Id: NotifyTest.cpp,v 1.1.6.2 2014/02/12 08:06:13 millayaraja Exp $  
 ****************************************************************************/
#ident "$Id: NotifyTest.cpp,v 1.1.6.2 2014/02/12 08:06:13 millayaraja Exp $"

#include <app.h>
#include <s6a_defines.h>
#include <s6a_stats.h>
#include <diameter/s6a_r10/Notify.h>

using namespace std;
using namespace its;
using namespace diameter;
using namespace diameter::base;
using namespace diameter::s6a_r10;

extern string sessIdString;
extern ITS_UINT sIdx;


enum NORFlagsValue
{
    Single_Registration_Indication = 1 << 0,
    SGSN_Area_Restricted   = 1 << 1,
    Ready_For_SM_From_SGSN = 1 << 2,
    UE_Reachable_From_MME  = 1 << 3,
    NOT_Flag_Reserver      = 1 << 4,
    UE_Reachable_From_SGSN = 1 << 5,
    Ready_For_SM_From_MME  = 1 << 6,
    Homogeneous_Support_Of_Voice_Over_PS_Sessions = 1 << 7
};



/*.implementation:public
 ****************************************************************************
 *  Purpose: This function sends the  Notify to the
 *      diameter stack. Create the  Notify object,
 *      populate the AVPs and send the message to the stack.
 *
 *  Input Parameters:
 *      Update choice.
 *
 *  Input/Output Parameters:
 *      None.
 *
 *  Output Parameters:
 *      None.
 *
 *  Return Value:
 *
 *  Notes:
 *
 *  See Also:
 ****************************************************************************/
int SendS6ANotifyRequest(bool readyForSM)
{
    // [1]. Create an instance of NotifyRequest.
    // Optional param error - whether this command carries an error indicator.
    // Optional param retrans - whether this is a retransmission.
    diameter::s6a_r10::NotifyRequest NOReq;

    // [2.1.1.x] Fixed AVP
    // [2.1.1.1] Set the Session-Id AVP.
    // The Session-Id AVP is used to identity a specific session
    // and must appear immediately following the Diameter header.
    // The AVP Code of Session-Id is 263
    // The data type of Session-Id is UTF8String.
    AllocateSession(S6A_INTERFACE_APP_ID, sIdx, sessIdString);
    NOReq.setSessionId(diameter::base::SessionId(sessIdString));

    NOReq.setSessionIndex(sIdx);

    // [2.1.2.x] Required AVPs
    // [2.1.2.2]Create a Auth-Session-State AVP object.
    // The Auth-Session-State AVP is specifies whether state is
    // maintained for a particular session.
    // The AVP Code of Auth-Session-State AVP is 277
    // The data type of Auth-Session-State AVP is Enumerated with
    // the following values
    // {
    //  STATE_MAINTAINED(0),
    //  NO_STATE_MAINTAINED(1)
    // }
    AuthSessionState authSessionState(AuthSessionState::NO_STATE_MAINTAINED);
    NOReq.setAuthSessionState(authSessionState);


    // [2.1.2.3] Set the Origin-Host AVP.
    // The Origin-Host AVP identifies the end-point that originated
    // the Diameter message.
    // The Origin-Host AVP must be present in all the Diameter messages.
    // This AVP SHOULD be placed as close to the Diameter header as possible.
    // The relay agents must not modify this AVP.
    // The AVP Code of Origin-Host AVP is 264
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginHost originHost(GetOriginHost());
    NOReq.setOriginHost(originHost);

    // [2.1.2.4] Set the Origin-Realm AVP.
    // The Origin-Realm AVP contains the realm of the originator of
    // any Diameter message and must be present in all the Diameter messages.
    // The AVP Code of Origin-Host AVP is 296.
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginRealm originRealm(GetOriginRealm());
    NOReq.setOriginRealm(originRealm);


    // [2.1.2.5] Set the Destination-Host AVP.
    // [2.2.2] Populate optional paramete "Destination-Host".
    // Set the Destination-Host AVP.
    // The data type of Destination-Host is DiameterIdentity
    // This AVP identifies the endpoint where  the Diameter message is destinated
    // This AVP SHOULD be placed as close to the Diameter header
    // The AVP Code of Destination-Host is 293.
    DestinationHost DestinationHost(GetDestinationHost());
    NOReq.setDestinationHost(DestinationHost);

    // [2.1.2.6] Set the Desti
    // The Destination-Realm AVP contains the realm, the message has
    // to be routed to.
    // The Destination-Realm AVP is used to perform message routing decisions.
    // The AVP Code of Destination-Realm AVP is 283
    // The data type of Destination-Realm AVP is DiameterIdentity.
     DestinationRealm destinationRealm(GetDestinationRealm());
     NOReq.setDestinationRealm(destinationRealm);

    // [2.1.2.7] Set the User-Name AVP.
    // The User-Name AVP (AVP Code 1) [RADIUS] is of type UTF8String, 
    // which contains the User-Name, in a format consistent with the NAI
    // specification [NAI]. The AVP Code of User-Name AVP is 1.
    // The data type of Usere-Name AVP is UTF8String.
    //UserName userName("sip:user@3gppnetwork.com");
    //ulReq.setUserName(userName);

    static long long imsi_number=singlton::arg[0];
        char temp[100]={'\0'},*end;
        char s[100]="@3gppntework.org";
        cout<<"IMSI BASED ROUTING"<<endl;
        sprintf(temp,"%lld%s",imsi_number,s);
        UserName userName((const char *)temp);
        NOReq.setUserName(userName);
        if(imsi_number!=singlton::arg[1])
                imsi_number++;
        else
                imsi_number=singlton::arg[0];

    if (readyForSM)
    {
        NOReq.setNORFlags(NORFlags(Ready_For_SM_From_MME));
        NOReq.setAlertReason(AlertReason(AlertReason::UE_MEMORY_AVAILABLE));
    }
    else
    {
        NOReq.setNORFlags(NORFlags(Single_Registration_Indication | 
                          UE_Reachable_From_MME |
                          SGSN_Area_Restricted));
    }


    // Application should not set this AVP.
    // NOReq.addProxyInfo(diameter::base::ProxyInfo(...));
    // NOReq.addRouteRecord(diameter::base::RouteRecord(...));
    // add more AVP as ANY values

    if (appMode == UNIT_TEST)
    cout<<NOReq;

    // Send object to stack.
    SendDiaAppMsg(NOReq, sIdx);

    S6AStats::GetS6AStats()->UpdateSendStats(NOReq.getCommandCode(), ITS_TRUE);
    return ITS_SUCCESS;
}


/*.implementation:public
 ****************************************************************************
 *  Purpose:
 *      This function sends the Answer to the Notify request
 *      made by the diameter stack. Create the NotifyAnswer
 *      object, populate the AVPs and send the message to the stack.
 *
 *  Input Parameters:
 *      const diameter::CommandImpl *impl: Pointer to the received Request
 *      Command, for which the Answer is to be sent.
 *
 *  Input/Output Parameters:
 *      None.
 *
 *  Output Parameters:
 *      None.
 *
 *  Return Value:
 *
 *  Notes:
 *
 *  See Also:
 ****************************************************************************/
int SendS6ANotifyAnswer(const CommandImpl * impl)
{
    // 1. Create an instance of NotifyAnswer class
    // The NotifyAnswer command is sent by a server
    // in response to the NotifyRequest command with
    // Command-Code = xxx and 'R' bit reset
    // Optional param error - whether this command carries an error indicator.
    // Optional param retrans - whether this is a retransmission.
    diameter::s6a_r10::NotifyAnswer NOAns;

    // Create an instance of NotifyRequest class to get host and
    // session details
    NotifyRequest NOReq(*impl);

    // Get AVP values from requested message
    const SessionId &sid1 = NOReq.getSessionId();
    const OriginHost &oh1 = NOReq.getOriginHost();
    const OriginRealm &or1 = NOReq.getOriginRealm();
    const AuthSessionState &authSessState = NOReq.getAuthSessionState();

    // 2. Create the instances of the classes describing the various
    // AVPs' present in an NotifyAnswer command and
    // pass them by value to the corresponding data members set
    // methods of the NotifyAnswer object created
    // in step 1.
    // [2.1.1] Set the Session-Id AVP.
    // The Session-Id AVP is used to identity a specific session
    // and must appear immediately following the Diameter header.
    // The AVP Code of Session-Id is 263
    // The data type of Session-Id is UTF8String.
    NOAns.setSessionId(sid1);
    NOAns.setSessionIndex(NOReq.getSessionIndex());    
 
    // [2.1.2.x] Required AVPs
    NOAns.setAuthSessionState(authSessState);
   
    // [2.2] Set the Result-Code AVP .
    // The Result-Code AVP indicates whether a particular request
    // was completed successfully or whether an error occurred
    // All Diameter answer messages must contain one Result-Code AVP
    // Diameter provides the following classes of error which
    // is represented by the Result-Code AVP:
    // 1xxx - Informational
    // 2xxx - Success
    // 3xxx - Protocol Errors
    // 4xxx - Transient Failures
    // 5xxx - Permanent Failure
    // The AVP Code of Result-Code AVP is 268.
    // The data type of Result-Code AVP is Unsigned32.
    // already ExperimentalResult is been set
    ResultCode resultCode(2001u);
    NOAns.setResultCode(resultCode);

    // Create the Experimental-Result AVP
    // The Experimental-Result AVP indicates whether a particular
    // vendor specific request was completed successfully or whether
    // an error occurred.
    // The AVP Code of Experimental-Result AVP is 297.
    // The data type of Experimental-Result AVP is Grouped Type.
    ExperimentalResult experimentalResult;

    // Set the Vendor-Id AVP in Experimental-Result AVP.
    // The Vendor-Id AVP in the Experimental-Result AVP indicates
    // whether a particular vendor-specific request was completed
    // successfully or whether an error occurred.
    VendorId vendorId1(VENDOR_ID);
    experimentalResult.setVendorId(vendorId1);

    // Set the Experimental-Result-Code AVP.
    // The Experimental-Result-Code AVP contains a vendor-assigned
    // value representing the result of processing the request.
    // The AVP Code of AVP is 298 .
    // The data type of AVP is Unsigned32.
    ExperimentalResultCode experimentalResultCode(2001u);
    experimentalResult.setExperimentalResultCode(experimentalResultCode);

    // [2.2] Set the Experimental-Result AVP.

    // This AVP is required incase of Result-Code AVP is missing.
    NOAns.setExperimentalResult(experimentalResult);

    // [2.1.1] Set the Origin-Host AVP.
    // The Origin-Host AVP identifies the end-point that originated
    // the Diameter message.
    // The Origin-Host AVP must be present in all the Diameter
    // messages.
    // This AVP SHOULD be placed as close to the Diameter header as
    // possible.
    // The relay agents must not modify this AVP.
    // The AVP Code of Origin-Host AVP is 264
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginHost originHost(GetOriginHost());
    NOAns.setOriginHost(originHost);

    // [2.1.1] set the Origin-Realm AVP.
    // The Origin-Realm AVP contains the realm of the originator of
    // any Diameter message and must be present in all the Diameter
    // messages.
    // The AVP Code of Origin-Host AVP is 296.
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginRealm originRealm(GetOriginRealm());
    NOAns.setOriginRealm(originRealm);

    // The data type of Proxy-Info is Grouped Type.
    // This AVP should not be set by Application.
    // NOAns.setProxyInfo(diameter::base::ProxyInfo(...));

    // Create the Route-Record AVP.
    // The identity added in the Route-Record AVP must be the same as
    // the one received in the Origin-Host AVP of the message.
    // The AVP Code of Route-Record AVP is 282
    // The data type of Route-Record is DiameterIdentity.
    // This AVP should not be set by Application.
    // NOAns.setRouteRecord(diameter::base::RouteRecord(...));
    // add more AVP as ANY values

    if (appMode == UNIT_TEST)
    cout<<NOAns;

    // Send Object to stack.
    SendDiaAppMsg(NOAns, sIdx);

    // Update stats.
    S6AStats::GetS6AStats()->UpdateSendStats(NOAns.getCommandCode(), ITS_FALSE);
    return ITS_SUCCESS;
}
